shotover/src/transforms/kafka/sink_cluster/mod.rs

```diff
@@ -825,32 +825,18 @@ routing message to a random node so that:
                     let destination = if let Some(partition) =
                         topic_meta.partitions.get(partition_index)
                     {
-                        if let Some(node) = self
-                            .nodes
-                            .iter_mut()
-                            .filter(|node| {
-                                partition
-                                    .shotover_rack_replica_nodes
-                                    .contains(&node.broker_id)
-                            })
-                            .choose(&mut self.rng)
-                        {
-                            node.broker_id
-                        } else {
-                            tracing::debug!(
-                                "Routing fetch request to replica outside of shotover's rack"
+                        // While technically kafka has some support for fetching from replicas, its quite weird.
+                        // See https://cwiki.apache.org/confluence/display/KAFKA/KIP-392%3A+Allow+consumers+to+fetch+from+closest+replica
+                        // We should never route to replica_nodes from the metadata response ourselves.
+                        // Instead, when its available, we can make use of preferred_read_replica field in the fetch response as an optimization.
+                        // However its always correct to route to the partition.leader_id which is what we do here.
+                        if partition.leader_id == -1 {
+                            let topic_name = Self::format_topic_name(&topic);
+                            tracing::warn!(
+                                "leader_id is unknown for topic {topic_name} at partition index {partition_index}"
                             );
-                            self.nodes
-                                .iter_mut()
-                                .filter(|node| {
-                                    partition
-                                        .external_rack_replica_nodes
-                                        .contains(&node.broker_id)
-                                })
-                                .choose(&mut self.rng)
-                                .unwrap()
-                                .broker_id
                         }
+                        partition.leader_id
```

Comment:
I'm not too familiar with kafka fetch requests, but should shotover route it to replica nodes if `partition.leader_id == -1` ?

Reply:
unfortunately not, from my understanding of https://cwiki.apache.org/confluence/display/KAFKA/KIP-392%3A+Allow+consumers+to+fetch+from+closest+replica only the leader should be used, in some circumstances however we can refer to `preferred_read_replica` field which may point at replicas but I've not attempted that in this PR.

Reply:
probably a silly question but what happens to the fetch requests when `partition.leader_id == -1`?

Reply:
The only time I've seen a leader_id of -1 is when the leader broker was killed and a new leader has not yet been found.

Any time shotover routes a request to broker -1 it routes the request to a random node.
Kafka will then give an appropriate error message for being unable to handle the request which we then return the client.
The client will then retry the request.